### PR TITLE
Validate PushConstants annotation and type

### PR DESCRIPTION
**Pull request by ale64bit**
Check that PushConstant OpVariable is identified by Block decoration and type is struct or array of structs.

### CODE REVIEW

**Comment by alan-baker** [comment]
Addresses #2005

File: source/val/validate_decorations.cpp [hunk 1]
```diff
@@ -795,6 +795,15 @@ spv_result_t CheckDecorationsOfBuffers(ValidationState_t& vstate) {
         MemberConstraints constraints;
         ComputeMemberConstraintsForStruct(&constraints, id, LayoutConstraints(),
                                           vstate);
+
+        if (push_constant && !hasDecoration(id, SpvDecorationBlock, vstate)) {
```
Comment (alan-baker) [inline source/val/validate_decorations.cpp]: This check should also be gated by `spvIsVulkanEnv(vstate.context()->target_env)`. It does not apply generally to SPIR-V.

File: source/val/validate_memory.cpp [hunk 2]
```diff
@@ -405,6 +405,17 @@ spv_result_t ValidateVariable(ValidationState_t& _, const Instruction* inst) {
                << "must be typed as OpTypeStruct, or an array of this type";
       }
     }
+
+    if (storage_class == SpvStorageClassPushConstant) {
```
Comment (alan-baker) [inline source/val/validate_memory.cpp]: Same here for the environment.

File: test/val/val_decoration_test.cpp [hunk 3]
```diff
@@ -2219,6 +2219,38 @@ TEST_F(ValidateDecorations,
           "rules: member 1 at offset 4 is not aligned to 16"));
 }
 
+TEST_F(ValidateDecorations, PushConstantMissingBlockVulkanBad) {
```
Comment (alan-baker) [inline test/val/val_decoration_test.cpp]: This test should succeed with a universal environment. Please modify this test to use the Vulkan environment and add a new test for the universal environment.

File: test/val/val_memory_test.cpp [hunk 4]
```diff
@@ -838,6 +838,34 @@ TEST_F(ValidateMemory, ArrayLenIndexNotPointerToStruct) {
           "The Struture's type in OpArrayLength <id> '12' must be a pointer to "
           "an OpTypeStruct.\n  %12 = OpArrayLength %uint %11 0\n"));
 }
+
+TEST_F(ValidateMemory, VulkanPushConstantNotStructBad) {
```
Comment (alan-baker) [inline test/val/val_memory_test.cpp]: Similar treatment here. Add a passing version in the universal environment.

File: source/val/validate_memory.cpp (hunk 2 above)
Comment (ale64bit) [reply source/val/validate_memory.cpp]: Isn't this already gated above?
https://github.com/KhronosGroup/SPIRV-Tools/blob/70fdd2e3a0bc71918a1d34a31584077279557890/source/val/validate_memory.cpp#L377

I added a comment and reorganized according to the spec, if that helps.

**Comment by ale64bit** [comment]
PTAL

**Comment by alan-baker** [comment]
Filed https://github.com/KhronosGroup/glslang/issues/1609 for the invalid glslang test this check exposes.

**Pull request merged**